### PR TITLE
Add snmp input plugin

### PR DESCRIPTION
**Pull request by titilambert**
This is a new PR of #453 rebased on 0.3.0

### CODE REVIEW

**Comment by sparrc** [comment]
@titilambert I don't quite understand how this is configured? Users can specify an OID file from snmptranslate, or they can input each OID into the config file?

File: plugins/inputs/snmp/snmp.go [hunk 1]
```diff
@@ -0,0 +1,413 @@
...
+	Get               []Data
+	Bulk              []Data
+	Snmptranslatefile string
+}
+
+type Host struct {
+	Address   string
+	Community string
+	// SNMP version. Default 2c
+	Version int
```
Comment (sparrc) [inline plugins/inputs/snmp/snmp.go]: shouldn't this be a string?

File: plugins/inputs/snmp/snmp.go [hunk 2]
```diff
@@ -0,0 +1,413 @@
...
+	// SNMP timeout, in seconds. 0 means no timeout
+	Timeout float64
+	// SNMP retries
+	Retries int
+	// Collect
+	Collect []string
```
Comment (sparrc) [inline plugins/inputs/snmp/snmp.go]: What does `Collect` do?

**Comment by TheFlyingCorpse** [comment]
How would I go forward with this to get a table with interfaces and their metrics with correlation?

Would it be possible to choose one value over another, depending if it is present or not? Thinking IF-MIB 's ifXTable (counter64) vs interfaces (counter32) issue on some appliances even to this day.

I have a lot of data I want to gather that is in the snmp table format so I think this is imporant. I dont neccesarily want all columns/rows from the table either, I want to have some way to choose which results to insert. Maybe both on a given alias/name or only these columns. Thoughts?

**Comment by titilambert** [comment]
Hello ! @TheFlyingCorpse I worked on an other SNMP project called SNMPbooster for Shinken (http://shinken-module-snmp-booster.readthedocs.org/en/latest/). I can try to adapt its concept to this plugin if you find it useful ...
Tell me what you think about.

**Comment by titilambert** [comment]
@sparrc I add some comments in sample config output.
BTW, I waiting the @TheFlyingCorpse response, But I think this plugin could be merged now, And I could add feature when needs will be more clear.

**Comment by TheFlyingCorpse** [comment]
@titilambert - if you could provide examples for how you would do tables like interfaces/ifXtable that would be great and really what I would need to be able to start testing myself.
SNMPbooster seems a bit complex for something simple like polling 1-2000 nodes for SNMP and giving that data to a backend like influxdb. 
I was thinking of querying Influxdb with some tool to get a simple holtWinters or linear prediction to see if the values deviate too far in any way and show this in icinga or similar based on what value type it is (disk usage for example). Maybe Kapacitor could watch in the stream for this in the future :)

**Comment by titilambert** [comment]
@TheFlyingCorpse here some examples
If you want to get a specific OID (with SNMP get)
```
  [[plugins.snmp.get]]
    name = "sysuptime"
    oid = ".1.3.6.1.2.1.1.3.0"
    unit = "seconds" 
```

if you want to get a entire table (with SNMP bulk)
```
  [[plugins.snmp.bulk]]
    name = "ifspeed"
    max_repetition = 127
    oid = ".1.3.6.1.2.1.2.2.1.5"
```

Then set your hosts
```
  [[plugins.snmp.host]]
    address = "192.168.2.2:161"
    community = "public"
    version = 2
    timeout = 2.0
    collect = ["sysuptime" ,"ifspeed"]

  [[plugins.snmp.host]]
    address = "192.168.2.3:161"
    community = "public"
    version = 2
    timeout = 2.0
    collect = ["sysuptime" ]
```
The ```collect``` array define which oid you want to get for the host.

To get oid translated
```
[[plugins.snmp]]
  snmptranslatefile = "/tmp/oids.txt"
  [[plugins.snmp.host]]
    address = "192.168.2.3:161"
    community = "public"
    version = 2
    timeout = 2.0
    collect = ["sysuptime" ]

  [[plugins.snmp.get]]
    name = "sysuptime"
    oid = ".1.3.6.1.2.1.1.3.0"
    unit = "seconds"
    snmptranslate = true
```
where ```/tmp/oids.txt``` is generated with ```snmptranslate -m all -Tz -On | sed -e 's/"//g' > /tmp/oids.txt```

**Comment by titilambert** [comment]
@TheFlyingCorpse about the ifXtable example, unfortunately, you have to add each OID manually (or script generated) :/ 

```
  [[plugins.snmp.get]]
    name = "eth0_in_octets"
    oid = "iso.3.6.1.2.1.31.1.1.1.6.1"
    snmptranslate = true

  [[plugins.snmp.get]]
    name = "eth1_in_octets"
    oid = "iso.3.6.1.2.1.31.1.1.1.6.2"
    snmptranslate = true

  [[plugins.snmp.get]]
    name = "eth0_out_octets"
    oid = "iso.3.6.1.2.1.31.1.1.1.10.1"
    snmptranslate = true

  [[plugins.snmp.get]]
    name = "eth1_out_octets"
    oid = "iso.3.6.1.2.1.31.1.1.1.10.2"
    snmptranslate = true
```

**Comment by titilambert** [comment]
@TheFlyingCorpse about your first question, I'm agree with you to get something better with "metrics correlation". But, this feature need this plugin have to store some data between two checks... and I don't know how to do it. Maybe @sparrc can help me on this point ?

**Comment by titilambert** [comment]
@TheFlyingCorpse to be short, this is only the first version of this plugin, the next one will be much better ;)

**Comment by TheFlyingCorpse** [comment]
I'm not thinking correlation for past checks, thats up to the backend to do (counter32/64 etc). What I was hoping for was an example of how I could easily get ifXtable and select which columns (or rows by some string to an oid?), not adding all OIDs manually. What else are the SNMP MIB conversion process for?

**Comment by titilambert** [comment]
@TheFlyingCorpse SNMP conversion process, it just convert your mib files into a single file with lines like:
```
...
ifHCInOctets			1.3.6.1.2.1.31.1.1.1.6
ifHCInUcastPkts			1.3.6.1.2.1.31.1.1.1.7
ifHCInMulticastPkts			1.3.6.1.2.1.31.1.1.1.8
ifHCInBroadcastPkts			1.3.6.1.2.1.31.1.1.1.9
ifHCOutOctets			1.3.6.1.2.1.31.1.1.1.10
ifHCOutUcastPkts			1.3.6.1.2.1.31.1.1.1.11
ifHCOutMulticastPkts			1.3.6.1.2.1.31.1.1.1.12
ifHCOutBroadcastPkts			1.3.6.1.2.1.31.1.1.1.13
...
```
It use to send ```ifHCInOctets``` instead of ```1.3.6.1.2.1.31.1.1.1.6``` to influxDB ....

Can we discuss about your need on IRC ?

**Comment by TheFlyingCorpse** [comment]
@titilambert Sure, I see I didnt catch the full example earlier. I'm on IRC under this nickname on freenode.

**Comment by titilambert** [comment]
@sparrc After a discussion with @TheFlyingCorpse you can merge it, if you think this is good. I will release a new version of this plugin with more features later

**Comment by sparrc** [comment]
Thanks @titilambert, can you change the PR to be pushing to `master`? that way it will be more clear the actual changes inbound

**Pull request closed without merging**